stream: add `WatchStream::from_changes`, add testing, etc.

brodybits opened the pull request:
other minor updates:

- add a test case for existing `WatchStream::from`
- update existing WatchStream test function name
- apply minor documentation updates

resolves #5420

<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/tokio-rs/tokio/blob/master/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation, which requires special commands beyond `cargo fmt` and `cargo doc`.
-->

## Motivation

see #5420

It looks like `WatchStream::from` was not covered by documentation or integration test cases, adding one here as well.

Quick disclaimer that I am still quite new with Rust in contrast to my years in quite a few other major systems programming languages.

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

## Solution

- added __`WatchStream::from_changes`__, with implementation as suggested in #5420, along with a test case and __minor__ documentation updates

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

Darksonn commented on tokio-stream/tests/watch.rs:
```diff
@@ -27,3 +27,34 @@ async fn message_not_twice() {
+    assert_eq!(stream.next().await.unwrap(), "bye");
+}
+
+#[tokio::test]
+async fn watch_stream_new_on_change() {
+    let (tx, rx) = watch::channel("hello");
+
+    let mut stream = WatchStream::new_on_changed(rx);
+
+    let task = tokio::spawn(async move {
+        let value = stream.next().await.unwrap();
+        if value != "bye" {
+            panic!("received unexpected value: {:?}", value);
+        }
+    });
+
+    tx.send("bye").unwrap();
+
+    task.await.unwrap();
+}
```
Since tests run on the current-thread runtime, this test is actually _guaranteed_ to call `send(bye)` before it calls `stream.next()`. This means that the test would also pass when using `new`.

Please use our tokio-test utilities to instead assert that `next` returns pending before the send, and then assert that it returns `"bye"` after the send.

brodybits replied:
should be resolved by some rework of this test

brodybits commented on tokio-stream/tests/watch.rs:
```diff
@@ -27,3 +32,29 @@ async fn message_not_twice() {
+
+    let mut stream = WatchStream::new_on_changed(rx);
+
+    assert_pending!(Pin::new(&mut stream).poll_next(&mut Context::from_waker(&noop_waker())));
```
I think this is quite ugly ... any suggestions?

extremeandy replied:
It's not so bad if you split it up and do it once at the start of the test. Also you can use `poll_next_unpin` from `StreamExt`.

```rust
let waker = futures::task::noop_waker_ref();
let mut cx = std::task::Context::from_waker(&waker);
...
assert_pending!(stream.poll_next_unpin(&mut cx));
```

Suggest being consistent with use of `assert_pending` and `assert_ready` rather than mixing that with async too.

E.g. rather than

`assert_eq!(stream.next().await.unwrap(), "bye");`
you can do
`assert_ready!(stream.poll_next_unpin(&mut cx), Some("bye"));`

Then you can remove `#[tokio::test]` and just use `#[test]`

Darksonn replied:
If you used the utilities in the tokio-test crate, then you can simplify this by a lot. See e.g. `tokio/tests/sync_mutex_owned.rs` for an example.

brodybits replied:
done

extremeandy commented on tokio-stream/src/wrappers/watch.rs:
```diff
@@ -66,6 +67,13 @@ impl<T: 'static + Clone + Send + Sync> WatchStream<T> {
     }
+
+    /// Create a new `WatchStream` that waits for the value to be changed.
+    pub fn new_on_changed(rx: Receiver<T>) -> Self {
```
How about `new_skip_unchanged`? Or `new_skip_until_changed`?

Darksonn replied:
I think `new_on_changed` is fine.

Darksonn commented on tokio-stream/src/wrappers/watch.rs:
```diff
@@ -66,6 +67,13 @@ impl<T: 'static + Clone + Send + Sync> WatchStream<T> {
         }
     }
+
+    /// Create a new `WatchStream` that waits for the value to be changed.
+    pub fn new_on_changed(rx: Receiver<T>) -> Self {
```
I would like to see at least one example using this method in the documentation.

brodybits replied:
done

brodybits commented on tokio-stream/src/wrappers/watch.rs:
```diff
@@ -40,6 +41,31 @@ use tokio::sync::watch::error::RecvError;
+/// let (tx, rx) = watch::channel("hello");
+/// let mut rx = WatchStream::new_on_changed(rx);
+///
+/// // no output from rx is available at this point
+/// let from_poll = Pin::new(&mut rx).poll_next(&mut noop_context());
+/// assert_eq!(from_poll, Poll::Pending);
```
A little ugly; I think it would be ideal if we could add a utility function to help check whether a stream is pending or ready.

brodybits commented:
I have simplified the test with help from the tokio-test utility and have added an example to the documentation.

I have also aded a test to the documentation to assert that the watch stream is still pending until new data is available. If we keep this, it may be enough overlap to remove the test case I added.

brodybits commented:
I think this should be better now, please let me know if anything else is needed from my end to get this accepted and merged.

I think there is an overlap between the `watch_stream_new_on_change` test case and the documentation example that I have added with a poll test. I would prefer to keep the documentation with the poll test that I made but can remove the poll test from the documentation if needed.

Darksonn reviewed (commented):
Other than this comment, it looks good to me.

brodybits commented on tokio-stream/src/wrappers/watch.rs:
```diff
@@ -40,6 +41,29 @@ use tokio::sync::watch::error::RecvError;
+/// let (tx, rx) = watch::channel("hello");
+/// let mut rx = WatchStream::new_on_changed(rx);
+///
+/// // no output from rx is available at this point - let's check this:
+/// assert!(Pin::new(&mut rx).poll_next(&mut noop_context()).is_pending());
```
Works with help from `futures_util::future::FutureExt` thanks!!

Darksonn commented on tokio-stream/Cargo.toml:
```diff
@@ -37,6 +37,8 @@ async-stream = "0.3"
 parking_lot = "0.12.0"
 tokio-test = { path = "../tokio-test" }
 futures = { version = "0.3", default-features = false }
+futures-test = "0.3.26"
+futures-util = "0.3.26"
```
I don't think we need these. We already import `futures`.

Darksonn replied:
Note that `FutureExt` is available from the `futures` crate.

brodybits replied:
I removed `futures-test`. Unfortunately I still need `futures-util` to import `FutureExt` as I needed to get `now_or_never()` to work.

I noticed that `tokio/src/sync/watch.rs` is using `recv.changed().now_or_never()` from `crate::sync::watch::channel(0i32)` in some tests. I would like to investigate this a little further. Any pointers would still be highly appreciated as before.

brodybits replied:
`futures::future::FutureExt` worked for me thanks!

Darksonn reviewed (approved):
Thanks.
